### PR TITLE
PR 2/3 - feat(gamification): group user score history by week and calculate accumulated points (#296)

### PR DESCRIPTION
**Pull request by eariasvalor**
⚠️ Do not merge until PR 1/3 (#294) is merged first. This PR depends on `findByUserIdOrderByCreatedAtAsc` introduced there.

## Description

PR 2/3 of parent issue #281. Implements weekly aggregation logic in `UserScoreServiceImpl` and introduces the new `ScoresHistoryResponseDto` and `WeeklyPointsDto` to support the score history chart.

## Analysis of existing behavior

`UserScoreServiceImpl` previously returned a flat list of score entries via `getUserPointsHistory()` using `findByUserIdOrderByCreatedAtDesc`. No aggregation or accumulated total was calculated.

## What this PR adds

- `ScoresHistoryResponseDto` and `WeeklyPointsDto` — new DTOs for the weekly chart response
- `getUserScoresHistoryChart()` in `UserScoreService` and `UserScoreServiceImpl`
- Weekly aggregation logic grouped by ISO week (yyyy-Www)
- Progressive accumulated total per week (`accumulatedAtEnd`)
- `aggregationType` fixed to `"WEEKLY"`
- Defined behavior for empty history (`totalPoints: 0`, `history: []`)
- Unit tests for all aggregation scenarios (all passing, see screenshot below)

test-results-service-chart-aggregation-pr2

## Out of scope

- Filtering history by `activityType` — tracked in #311
- Configurable `aggregationType` (daily, monthly) — tracked in #310
- Controller changes — covered in PR 3/3

## Known limitations

`givenMixedSourcesInSameWeek_whenGetUserScoresHistoryChart_thenAllPointsAreSummedTogether` currently simulates multiple point sources using distinct `challengeId` values, as `activityType` is not yet available in `UserScoreDocument`. `@SuppressWarnings("java:S2699")` added temporarily. This test must be extended once #275 and #276 are merged.

## Technical debt

- `aggregationType` hardcoded to `"WEEKLY"` — tracked in #310
- `activityType` filtering not yet supported — tracked in #311

## Changes after initial review

- Renamed `ScoresHistoryChartDto` to `ScoresHistoryResponseDto` to decouple the DTO name from a specific use case

## Self-checklist

- [x] I tested my changes locally and verified they work as expected
- [x] The PR has a clear and focused purpose (only one feature, fix or refactor)
- [x] Title, description, and changelog (if needed) are filled in correctly
- [x] There are no leftover merge conflicts or unrelated changes from previous branches
- [x] All automated checks (like SonarCloud) have passed
- [x] I responded to all previous review comments and only resolved those I truly addressed
- [x] I ran SonarLint locally on the modified files and confirmed there are no warnings or errors

### CODE REVIEW

**Review by carlosPc1987 — CHANGES_REQUESTED** [review]
Looks good overall and test coverage is strong. I found one important issue: weekly grouping currently uses DateTimeFormatter.ofPattern("YYYY-'W'ww").withLocale(Locale.getDefault()), which makes week calculation environment/locale-dependent and not strictly ISO. Since this feature is explicitly ISO week based (Mon–Sun), could we switch to WeekFields.ISO when building the period key (e.g. weekBasedYear + weekOfWeekBasedYear)?
Also, ResolverStyle.STRICT is not needed here because we are formatting, not parsing.

**Comment by eariasvalor** [comment]
switched to `WeekFields.ISO` for deterministic ISO week calculation, removing the locale dependency and the `DateTimeFormatter` entirely. `ResolverStyle.STRICT` also removed as it was only relevant for parsing, not formatting.

**Review by carlosPc1987 — APPROVED** [review]
Thanks for the update. WeekFields.ISO addresses the locale-dependent week grouping concern and removing ResolverStyle.STRICT is correct here. LGTM.

**Review by JordiCasas87 — APPROVED** [review]
Looks good 👍

No blocking issues from my side.

File: itachallenge-challenge/src/test/java/com/itachallenge/gamification/service/UserScoreServiceImplTest.java [hunk 1]
```diff
@@ -133,4 +130,114 @@ void givenScoresWithNullPoints_whenGetUserPointsHistory_thenTotalPointsIgnoresNu
                 .expectNextMatches(response -> response.getTotalPoints() == 10)
                 .verifyComplete();
     }
+
+    @Test
+    void givenScoresInSameWeek_whenGetUserScoresHistoryChart_thenWeeklyPointsEarnedIsCorrect() {
+        LocalDateTime week1Day1 = LocalDateTime.of(2024, 3, 4, 10, 0);
```
Comment (kenan-it-academy) [inline itachallenge-challenge/src/test/java/com/itachallenge/gamification/service/UserScoreServiceImplTest.java]: These local variables should be extracted and should have clearer names (for example `week1monday`).

File: itachallenge-challenge/src/test/java/com/itachallenge/gamification/service/UserScoreServiceImplTest.java [hunk 2]
```diff
@@ -133,4 +130,114 @@ void givenScoresWithNullPoints_whenGetUserPointsHistory_thenTotalPointsIgnoresNu
...
+        StepVerifier.create(userScoreService.getUserScoresHistoryChart(userId))
+                .expectNextMatches(response ->
+                        response.getHistory().size() == 1 &&
+                                response.getHistory().getFirst().getPointsEarned() == 150)
+                .verifyComplete();
+    }
+
+    @Test
+    void givenScoresAcrossMultipleWeeks_whenGetUserScoresHistoryChart_thenAccumulatedAtEndIsProgressiveSum() {
+        LocalDateTime week10 = LocalDateTime.of(2024, 3, 4, 10, 0);
```
Comment (kenan-it-academy) [inline itachallenge-challenge/src/test/java/com/itachallenge/gamification/service/UserScoreServiceImplTest.java]: Same with these and all following. No need to keep restating the same valuables (and please don't change naming schemes lol)

File: itachallenge-challenge/src/test/java/com/itachallenge/gamification/service/UserScoreServiceImplTest.java [hunk 3]
```diff
@@ -133,4 +130,114 @@ void givenScoresWithNullPoints_whenGetUserPointsHistory_thenTotalPointsIgnoresNu
...
+
+        when(userScoreRepository.findByUserIdOrderByCreatedAtAsc(userId))
+                .thenReturn(Flux.just(source1, source2));
+
+        StepVerifier.create(userScoreService.getUserScoresHistoryChart(userId))
+                .expectNextMatches(response ->
+                        response.getHistory().size() == 1 &&
+                                response.getHistory().getFirst().getPointsEarned() == 90)
+                .verifyComplete();
+    }
```
Comment (kenan-it-academy) [inline itachallenge-challenge/src/test/java/com/itachallenge/gamification/service/UserScoreServiceImplTest.java:232]: Please add a test that checks that the week changes at the correct point (specifically a week ends on Sunday and starts on Monday)

**Review by kenan-it-academy — CHANGES_REQUESTED** [review]

**Comment by eariasvalor** [comment]
Hi @kenan-it-academy, thanks for the feedback, in UserScoreServiceImplTest I have extracted all common variables from the individual tests and declared them as static final with proper names, everything should look OK now. Also, I have added a test that checks boundaries between weeks.

**Review by carlosPc1987 — APPROVED** [review]
Thanks for the update, @eariasvalor.
I can see the requested cleanup was applied (clearer/extracted week constants) and the new ISO week-boundary test (Sunday -> Monday) was added, which addresses the review concern.
From my side this looks good

File: itachallenge-challenge/src/test/java/com/itachallenge/gamification/service/UserScoreServiceImplTest.java (hunk 1 above)
Comment (eariasvalor) [reply itachallenge-challenge/src/test/java/com/itachallenge/gamification/service/UserScoreServiceImplTest.java]: Local variables extracted with legit names :)

**Review by Andrescalvo22 — APPROVED** [review]
Looks really solid overall 👍

The weekly aggregation logic is clear and well implemented. Using ISO week calculation together with a LinkedHashMap to preserve order works very well for the accumulated totals. The separation between flat history and chart aggregation also keeps the service clean and easy to follow.

The test coverage is quite complete, especially around edge cases like empty history, multiple weeks, and accumulation behavior. Also, good call documenting the limitation around activityType and using @SuppressWarnings with context.

Just a couple of minor (non-blocking) suggestions:

* The aggregationType being hardcoded as "WEEKLY" could be extracted to a constant (or future enum) to avoid magic strings.
* The null check with `continue` works fine, but in the future it might be worth considering filtering earlier or adding some logging for skipped records.

Aside from that, everything looks consistent and aligned with the expected behavior.

Good work, happy to approve 👍

**Comment by eariasvalor** [comment]
@Andrescalvo22, thanks for the thorough review! Extracted `"WEEKLY"` to a `AGGREGATION_TYPE_WEEKLY` constant to avoid the magic string.

Regarding the null check logging — agreed it could be useful in production, but I'd rather keep it simple for now and track it as technical debt to avoid adding noise before we have a clearer picture of how often this case actually occurs.

**Comment by eariasvalor** [comment]
MUserScoreServiceImplTest refactored including a helper method to build UserScoreDocument to reduce boilerplate in test setup. Also aligned `userId` consistently across all test documents.

**Comment by carlosPc1987** [comment]
Looks good to me, the ISO week boundary test (Sunday -> Monday) is now covered,
and the extracted/renamed constants make the test suite much clearer and easier to maintain. 
Good Job Esther.

**Review by kenan-it-academy — APPROVED** [review]
Seems good now!
